Several updates to the Access Rules article (#597)
gristlabs/grist-help: help/en/docs/access-rules.md

![Access rules](images/access-rules/access-rules-example.png)

## Enabling access rules

> Without any access ruless, every collaborator on a document can see all data in it, and every

Reviewer: ```suggestion
With access rules disabled, every collaborator on a document can see all its data, and every
```

Editor or Owner may change anything about the documement's data, structure, logic, or layout. In
other words, Owners and Editors are equally powerful within a document, with the ability to create
or delete tables or columns, write formulas, reorganize pages, and so on.

Enabling access rules changes this default to allow only Owners to change the document's
structure, and denying this permission to Editors.

> Click "Enable access rules" to add access rules to your document:

Reviewer: In the screenshot the button is capitalized. Do you know if this is changing? Otherwise let's just match.

![Enable access rules](images/access-rules/enable-access-rules.png)

You'll see the explanation of change: that Editors will no longer be able to change the structure
of the document or edit formulas, and that only Owners will be able to copy or download the
document. Once you click "Continue", you'll see the configuration for granular access rules.

Note that the presence of the enabled `Save` button is telling us that the changes (in this case
the enabling of access rules) have not been applied yet. Click `Save` for them to take effect.

To revert back to the default state, click the "Disable access rules" button at the bottom of the
page. If you don't see one, delete any other rules you see using the
"Trash" icon (<span class="grist-icon" style="--icon: var(--icon-Remove)"></span>);
the "Disable access rules" button will appear once there are no deletable rules.

> ![Disable access rules](images/access-rules/disable-access-rules.png)

Reviewer: I would consider removing the arrow to "Disable Access Rules" and moving the image after "...you’ll see the configuration for granular access rules." I think that makes much more sense visually and the "disable" description is detailed enough.

## Default rules

> To see the access rules for a document, visit its access rules page by

Reviewer: ```suggestion
To enable and view the access rules for a document, visit its access rules page by
```

These rules say, in summary, that Owners and Editors can do anything within
the document, that Viewers can only read the document, and everyone
else is forbidden all access.  These rules cannot be modified since they reflect the underlying enforcement of basic roles.

However, other rules can be added above them to restrict access in more granular ways, using the

> "+" button circled in the screenshot. Default rules in this section are used as the fallback for

Reviewer: ```suggestion
"+" button circled in the above screenshot. Default rules in this section are used as the fallback for
```

![access-rules-default-owner-access](images/access-rules/access-rules-default-owner-access.png)

## Special rules

Below "Default rules", you'll find several checkboxes that control document behavior outside of
viewing and editing data.

Each special rule has a checkbox to toggle it, and also an expander (`>`) that shows the complete
conditions that implement that special rule. The conditions may be edited if you'd like to toggle
that behavior for particular subsets of users (e.g. determined using
[user attributes](#user-attribute-tables)).

### Allow Editors to edit structure {: #lock-down-structure data-toc-label='Structure edit' }

The first is "Allow Editors to edit structure". This restores the default behavior of a document
without access rules. In fact, when that's the only checked rule (and no other rules for any

> tables), it's equivalent to disabling access rules altogether.

Reviewer: Don't know if we need the transition. Also, the use of "default behaviour" needs explanation (or a new term). Do we need to mention this? I don't want to get into the weeds about how "this goes back to the behaviour prior to the new enable/disable flow...". What do you think?

```suggestion
This rule restores the default behavior of a document without access rules. In fact, when this is the only checked rule (and there are no table rules), it's equivalent to disabling access rules altogether.
```

This rule controls a document-wide permission called the "structure permission", represented by
`S` if you expand it:

![Access rules](images/access-rules/access-rules-lock-structure.png)

**Important.** This structure permission (`S`) is very powerful. It allows creating or changing
formulas. Since formula calculations are not limited by access control rules, a determined user
could use them to retrieve any data from a document. In other words, having the "edit structure"

> permission makes it possible to circumvent other rules that prevent access to data.

Reviewer: ```suggestion
**Important:** The structure permission (`S`) is very powerful, as it allows for creating or changing
formulas. Since formula calculations are not limited by access rules, a determined user
could use them to retrieve any data from a document. In other words, having the structure
permission makes it possible to circumvent other rules that prevent access to data.
```

### Allow everyone to view access rules {: data-toc-label='View access rules' }

By default, only Owners may view access rules. This rule allows every collaborator to view them.
(It cannot allow non-Owners to change access rules, only view.)

> ![Special rules](images/access-rules/access-rules-special-rules.png)

Reviewer: I'd move this under the initial "Special rules" heading.

(such as `$Person.Email` or `rec.Person.Email`) are _not_ supported in access rule conditions.
Instead, in rule conditions, a reference value is a numerical ID of the referenced record.

You may still use such records in comparisons. For example:

-   **Example 1.** Let's imagine that table `Orders` has a column `Purchaser` that's a reference to the `Team`
    table and represents a person who placed the order. And let's say you have a user attribute
    `user.Team`, configured as above.

    Then you may compare them using the condition `rec.Purchaser == user.Team.id`{: .formula}.
    Here, `rec.Purchaser` is a numeric identifier of a record in table `Team`, while `user.Team` is
    such a record, and `user.Team.id` is its identifier (using the built-in `id` field).

-   **Example 2.** Let's say now that both tables `Orders` and `Team` each have a column named
    `Department`, which is a reference to the `Departments` table.

    To compare these, use this condition: `rec.Department == user.Team.Department`{: .formula}.

    Here, both `rec.Department` and `user.Team.Department` are numeric identifiers of records in

>     table `Departments`.

Reviewer: This section really needs style updating, and I might try to condense a bit upon my review. The additional context you added is helpful, though!

You can also use normal formulas in your actual table to bring in referenced data as additional
columns. E.g. if you add a column in table `Orders` named `Role` with the formula

> `$Purchaser.Role`, then you could use `rec.Role` in rule conditions for table `Orders`.

Reviewer: Maybe preface with "As a workaround," or something? This seems a bit suboptimal but I've often found myself wanting to create such helper columns.
